src/librustc_trans/cabi_wasm32.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use abi::{FnType, ArgType};
+use context::CodegenCx;
+
+fn classify_ret_ty<'a, 'tcx>(_cx: &CodegenCx<'a, 'tcx>, ret: &mut ArgType<'tcx>) {
+    ret.extend_integer_width_to(32);
+}
```

Comment:
This doesn't do the aggregate to indirect conversion anymore that the asm.js C ABI did: https://github.com/alexcrichton/rust/blob/74f5dd07cff8dba348f91cfef4df5242ab33a311/src/librustc_trans/cabi_asmjs.rs#L20-L33

However looking at how the C ABI for wasm is implemented in clang, it seems to still do it: https://github.com/llvm-mirror/clang/blob/master/lib/CodeGen/TargetInfo.cpp#L759-L775

Reply:
Ah yeah TBH I have very little clue what ABI-related code is doing...

Seems fine though for us to match clang! Until there's an "official ABI" though this'll probably be a game of cat and mouse.

Reply:
"Seems fine though for us to match clang!"
I'm a bit confused by that wording. What I'm saying is that we don't match clang anymore with this PR, so shouldn't we include the aggregate code as well to properly match it again?

Reply:
Er sorry, what I mean is that it seems fine to add. It won't have any effect today anyway as you can't intermingle wasm32-unknown-unknown and C yet so I'm hoping that we can fix this with a test in the future. (I'm also hesitant to add things that I don't personally understand in my own PRs)

Reply:
Well it also affects everyone who is calling an exported function from JS / a wasm environment. However I guess we don't really give any guarantees how structs are represented there anyway, so I guess it's fine to wait on the C ABI settling.

Reply:
Yeah with a lack of a spec about an ABI returning any sort of aggregate or non-integer/float is sort of "best effort" today anyway